File: hyperLiquid/src/helpers/marketPrice.helpers.ts
import * as hl from "@nktkas/hyperliquid";
import { AppError } from "./errorHandler";
import { HTTP } from "../constants/http";

/**
 * Get current market price (best bid/ask) for near-market execution
 * @param infoClient - HyperLiquid info client
 * @param symbol - Trading symbol (e.g., "BTC")
 * @param isBuy - True for buy orders, false for sell orders
 * @returns Adjusted market price for quick execution
 */
export async function getMarketPrice(
    infoClient: hl.InfoClient,
    symbol: string,
    isBuy: boolean
): Promise<number> {
    const allMids = await infoClient.allMids();
    const marketPrice = parseFloat(allMids[symbol] || "0");

    if (!marketPrice || isNaN(marketPrice)) {
        throw new AppError(`Unable to fetch market price for ${symbol}`, HTTP.BAD_REQUEST);
    }

    // For buy orders, price slightly above mid; for sell orders, slightly below
    // This ensures quick execution while avoiding market order restrictions
    const priceAdjustment = isBuy ? 1.001 : 0.999;
    return marketPrice * priceAdjustment;
}

/**
 * Format price according to HyperLiquid requirements
 * @param price - Raw price to format
 * @param szDecimals - Size decimals from asset info
 * @returns Formatted price string with proper decimal places
 */
export function formatPriceForOrder(price: number, szDecimals: number): string {
    const MAX_DECIMALS = 5;
    const pxDecimals = Math.max(0, MAX_DECIMALS - Math.floor(Math.log10(price)) - 1);
    return price.toFixed(Math.min(pxDecimals, szDecimals));
}

/**
 * Get and format market price in one step (convenience function)
 * @param infoClient - HyperLiquid info client
 * @param symbol - Trading symbol (e.g., "BTC")
 * @param isBuy - True for buy orders, false for sell orders
 * @param szDecimals - Size decimals from asset info
 * @returns Formatted market price ready for order placement
 */
export async function getFormattedMarketPrice(
    infoClient: hl.InfoClient,
    symbol: string,
    isBuy: boolean,
    szDecimals: number
): Promise<string> {
    const marketPrice = await getMarketPrice(infoClient, symbol, isBuy);
    return formatPriceForOrder(marketPrice, szDecimals);
}

File: src/hyperliquid_python/helpers/market_price_helpers.py
import math

def get_market_price(info_client, symbol: str, is_buy: bool) -> float:
    all_mids = info_client.all_mids()
    market_price = float(all_mids.get(symbol, 0))
    if not market_price:
        raise ValueError(f"Unable to fetch market price for {symbol}")

    price_adjustment = 1.001 if is_buy else 0.999
    return market_price * price_adjustment

def format_price_for_order(price: float, sz_decimals: int) -> str:
    # Max 5 significant figures
    max_decimals = 5
    if price <= 0:
        return "0"
    px_decimals = max(0, max_decimals - math.floor(math.log10(price)) - 1)
    # szDecimals is actually max decimals allowed sometimes, but hyperliquid API usually just uses 5 sig figs.
    # The ts code uses Math.min(pxDecimals, szDecimals) but let's just use 5 sig figs:
    formatted = f"{price:.{px_decimals}f}"
    return formatted

def get_formatted_market_price(info_client, symbol: str, is_buy: bool, sz_decimals: int) -> str:
    market_price = get_market_price(info_client, symbol, is_buy)
    return format_price_for_order(market_price, sz_decimals)
